File: src/jgmenu-config.py
#!/usr/bin/env python3
#
# Copyright (C) Johan Malm 2018
#

""" create/modify/query jgmenu config files """

import argparse
import sys
import os

DEFAULT_CONFIG_FILE = "~/.config/jgmenu/jgmenurc"

def jgmenurc():
    """ return dict with config file keys and values """
    keys_and_values = {
        "stay_alive": "1",
        "hide_on_startup": "0",
        "csv_cmd": "pmenu",
        "tint2_look": "1",
        "at_pointer": "0",
        "edge_snap_x": "30",
        "terminal_exec": "x-terminal-emulator",
        "terminal_args": "-e",
        "monitor": "0",
        "hover_delay": "100",
        "hide_back_items": "1",
        "columns": "1",
        "tabs": "120",
        "menu_margin_x": "0",
        "menu_margin_y": "0",
        "menu_width": "200",
        "menu_height_min": "0",
        "menu_height_max": "0",
        "menu_height_mode": "static",
        "menu_padding_top": "5",
        "menu_padding_right": "5",
        "menu_padding_bottom": "5",
        "menu_padding_left": "5",
        "menu_radius": "1",
        "menu_border": "0",
        "menu_halign": "left",
        "menu_valign": "bottom",
        "sub_spacing": "1",
        "sub_padding_top": "auto",
        "sub_padding_right": "auto",
        "sub_padding_bottom": "auto",
        "sub_padding_left": "auto",
        "sub_hover_action": "1",
        "item_margin_x": "3",
        "item_margin_y": "3",
        "item_height": "25",
        "item_padding_x": "4",
        "item_radius": "1",
        "item_border": "0",
        "item_halign": "left",
        "sep_height": "5",
        "sep_halign": "left",
        "sep_markup": "",
        "font": "",
        "font_fallback": "xtg",
        "icon_size": "22",
        "icon_text_spacing": "10",
        "icon_theme": "",
        "icon_theme_fallback": "xtg",
        "arrow_string": "▸",
        "arrow_width": "15",
        "color_menu_bg": "#000000 85",
        "color_menu_border": "#eeeeee 8",
        "color_norm_bg": "#000000 00",
        "color_norm_fg": "#eeeeee 100",
        "color_sel_bg": "#ffffff 20",
        "color_sel_fg": "#eeeeee 100",
        "color_sel_border": "#eeeeee 8",
        "color_sep_fg": "#ffffff 20",
        "color_scroll_ind": "#eeeeee 40",
        "csv_name_format": "%n (%g)",
        "csv_single_window": "0",
        "csv_no_dirs": "0",
        "csv_i18n": ""
        }
    return keys_and_values

def die(msg):
    """ exit with message """
    sys.stderr.write("fatal: {}\n".format(msg))
    sys.exit(1)

def get_key(line):
    """ return key from "key = value" string """
    s = ''
    if '=' in line:
        s = line.split('=', 1)[0].strip()
        if s[0] == '#':
            s = s[1:]
    return s

def get_keys(lines):
    """ extract keys from config file """
    keys = []
    for line in lines:
        s = get_key(line)
        if s and s != '':
            keys.append(s)
    return keys

def get_missing_keys(conf_lines):
    """ compare config file with built-in dict and return missing keys """
    conf_keys = get_keys(conf_lines)
    return [item for item in jgmenurc() if item not in set(conf_keys)]

def resolve(filename):
    """ expand ~ and $foo """
    if filename is None or filename == '':
        die("filename must be specified")
    filename = os.path.expanduser(filename)
    filename = os.path.expandvars(filename)
    return filename

def amend(conf_filename, isdryrun):
    """ amend config-file with missing items """
    conf_filename = resolve(conf_filename)
    if not os.path.exists(conf_filename):
        die("file '{}' does not exist".format(conf_filename))
    with open(conf_filename, "r+") as f:
        conf_lines = f.read().split("\n")

    missing_keys = get_missing_keys(conf_lines)
    if not missing_keys:
        print("info: there are no missing items in {}".format(conf_filename))
        return

    if isdryrun:
        print("info: missing keys {}".format(missing_keys))
        return

    print("info: add keys {}".format(missing_keys))
    for key in missing_keys:
        conf_lines.append("#{} = {}".format(key, jgmenurc()[key]))

    # write config file
    with open(conf_filename, "w") as f:
        for line in conf_lines:
            f.write("{}\n".format(line))

def create(conf_filename):
    """ create new config file """
    conf_filename = resolve(conf_filename)
    if os.path.exists(conf_filename):
        die("file '{}' already exists".format(conf_filename))
    l = []
    d = jgmenurc()
    for key in d:
        l.append("#{} = {}".format(key, d[key]))
    with open(conf_filename, "w") as f:
        for line in l:
            f.write("{}\n".format(line))

def main():
    """ script main program """
    parser = argparse.ArgumentParser(prog="jgmenu_run config")
    subparsers = parser.add_subparsers(dest="command", help="commands")
    subparsers.required = True
    create_parser = subparsers.add_parser("create", help="create config-file")
    create_parser.add_argument("--file", default=DEFAULT_CONFIG_FILE,
                               help="specify config file", metavar="FILE")
    amend_parser = subparsers.add_parser("amend", help="amend config-file \
                                         with missing items")
    amend_parser.add_argument("--file", metavar="FILE", help="specify config \
                              file", default=DEFAULT_CONFIG_FILE)
    amend_parser.add_argument("--dryrun", action="store_true",
                              help="list missing items, but do not amend file")
    args = parser.parse_args()
    if args.command == "create":
        create(args.file)
    elif args.command == "amend":
        amend(args.file, args.dryrun)
    else:
        parser.print_help(sys.stderr)

if __name__ == '__main__':
    main()

File: src/jgmenu-config.c
/*
 * jgmenu-config.c
 *
 * Copyright (C) Johan Malm 2019
 *
 * Create or amend config file.
 */

#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <stdarg.h>

#include "util.h"
#include "sbuf.h"
#include "compat.h"
#include "set.h"
#include "spawn.h"

struct option {
	const char *key;
	const char *value;
};

static struct option options[] = {
	{ "verbosity", "0" },
	{ "stay_alive", "1" },
	{ "hide_on_startup", "0" },
	{ "csv_cmd", "pmenu" },
	{ "tint2_look", "0" },
	{ "position_mode", "fixed" },
	{ "edge_snap_x", "30" },
	{ "terminal_exec", "x-terminal-emulator" },
	{ "terminal_args", "-e" },
	{ "monitor", "0" },
	{ "hover_delay", "100" },
	{ "hide_back_items", "1" },
	{ "columns", "1" },
	{ "tabs", "120" },
	{ "menu_margin_x", "0" },
	{ "menu_margin_y", "0" },
	{ "menu_width", "200" },
	{ "menu_height_min", "0" },
	{ "menu_height_max", "0" },
	{ "menu_height_mode", "static" },
	{ "menu_padding_top", "5" },
	{ "menu_padding_right", "5" },
	{ "menu_padding_bottom", "5" },
	{ "menu_padding_left", "5" },
	{ "menu_radius", "1" },
	{ "menu_border", "0" },
	{ "menu_halign", "left" },
	{ "menu_valign", "bottom" },
	{ "sub_spacing", "1" },
	{ "sub_padding_top", "auto" },
	{ "sub_padding_right", "auto" },
	{ "sub_padding_bottom", "auto" },
	{ "sub_padding_left", "auto" },
	{ "sub_hover_action", "1" },
	{ "item_margin_x", "3" },
	{ "item_margin_y", "3" },
	{ "item_height", "25" },
	{ "item_padding_x", "4" },
	{ "item_radius", "1" },
	{ "item_border", "0" },
	{ "item_halign", "left" },
	{ "sep_height", "5" },
	{ "sep_halign", "left" },
	{ "sep_markup", "" },
	{ "font", "" },
	{ "font_fallback", "xtg" },
	{ "icon_size", "22" },
	{ "icon_text_spacing", "10" },
	{ "icon_theme", "" },
	{ "icon_theme_fallback", "xtg" },
	{ "arrow_string", "▸" },
	{ "arrow_width", "15" },
	{ "color_menu_bg", "#000000 100" },
	{ "color_menu_border", "#eeeeee 8" },
	{ "color_norm_bg", "#000000 00" },
	{ "color_norm_fg", "#eeeeee 100" },
	{ "color_sel_bg", "#ffffff 20" },
	{ "color_sel_fg", "#eeeeee 100" },
	{ "color_sel_border", "#eeeeee 8" },
	{ "color_sep_fg", "#ffffff 20" },
	{ "color_scroll_ind", "#eeeeee 40" },
	{ "color_title_fg", "#eeeeee 50" },
	{ "color_title_bg", "#000000 0" },
	{ "color_title_border", "#000000 0" },
	{ "csv_name_format", "%n (%g)" },
	{ "csv_single_window", "0" },
	{ "csv_no_dirs", "0" },
	{ "csv_i18n", "" },
	{ NULL, NULL },
};

static const char config_usage[] =
"Usage: jgmenu_run config [options]\n"
"Creates or amend config file\n"
"Options:\n"
"    -a <file>      amend config file with missing items\n"
"    -c             write default config file to stdout\n"
"    -k <key>       specify key (needed for -s)\n"
"    -s <file>      set key/value pair (add if does not exist)\n"
"    -v <value>     specify value (needed for -s)\n"
"    -h             display this message\n";

static void usage(void)
{
	printf("%s", config_usage);
	exit(0);
}

static void create(void)
{
	struct option *o;

	for (o = options; o->key; o++)
		printf("# %s = %s\n", o->key, o->value);
}

static void check_file(struct sbuf *f, const char *filename)
{
	struct stat sb;

	if (!filename)
		die("no filename specified");
	sbuf_cpy(f, filename);
	sbuf_expand_tilde(f);

	/* if file does not exist, we create it */
	if (stat(f->buf, &sb)) {
		static const char * const command[] = { "jgmenu_run", "init",
							NULL };

		spawn_sync(command);
	}
}

static void amend(const char *filename)
{
	struct sbuf f;
	struct option *o;

	sbuf_init(&f);
	check_file(&f, filename);
	set_read(f.buf);
	for (o = options; o->key; o++) {
		if (!set_key_exists(o->key)) {
			info("add '%s = %s'", o->key, o->value);
			set_set(o->key, o->value, 1);
		}
	}
	set_write(f.buf);
	xfree(f.buf);
}

static void set_key_value_pair(const char *filename, const char *key,
			       const char *value)
{
	struct sbuf f;

	sbuf_init(&f);
	check_file(&f, filename);
	set_read(f.buf);

	/*
	 * Aborting at this point if the key/value pair is already set
	 * correctly, avoid updating the config file and thereby indirectly
	 * avoids restarting jgmenu too.
	 */
	if (set_is_already_set_correctly(key, value))
		goto out;
	set_set(key, value, 0);
	set_write(f.buf);
out:
	xfree(f.buf);
}

int main(int argc, char **argv)
{
	enum action { AMEND, CREATE, SET, UNKNOWN };
	enum action action = UNKNOWN;
	char *key = NULL, *value = NULL, *filename = NULL;
	int i;

	if (argc < 2)
		usage();
	for (i = 1; i < argc; i++) {
		const char *arg = argv[i];

		if (*arg == '-') {
			switch (arg[1]) {
			case 'a':
				action = AMEND;
				filename = argv[i + 1];
				++i;
				continue;
			case 'c':
				action = CREATE;
				continue;
			case 'h':
				usage();
				continue;
			case 'k':
				key = argv[i + 1];
				++i;
				continue;
			case 's':
				action = SET;
				filename = argv[i + 1];
				++i;
				continue;
			case 'v':
				value = argv[i + 1];
				++i;
				continue;
			default:
				die("unknown argument '%s'", arg);
			}
		}
		die("unknown argument '%s'", arg);
	}

	switch (action) {
	case AMEND:
		amend(filename);
		break;
	case CREATE:
		create();
		break;
	case SET:
		if (!key || !value)
			die("'key' and 'value' need to be specified");
		set_key_value_pair(filename, key, value);
		break;
	default:
		fprintf(stderr, "info: use -a, -c or -s\n");
		break;
	}

	return 0;
}
